File: backend/app/schemas/analytics.py
from datetime import date
from typing import Any, Optional

from pydantic import BaseModel


class MonthlySummary(BaseModel):
    month: str
    income: float
    expenses: float
    savings: float
    net: float


class CategoryBreakdown(BaseModel):
    category_id: Optional[str] = None
    category_name: str
    total: float
    percentage: float
    color: str


class MonthlyTrend(BaseModel):
    month: str
    income: float
    expenses: float


class ForecastOut(BaseModel):
    month: date
    predictions: dict[str, Any]
    total_predicted: Optional[float] = None
    confidence: Optional[str] = None
    generated_at: Optional[Any] = None

File: frontend/lib/core/models/analytics.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'analytics.freezed.dart';
part 'analytics.g.dart';

@freezed
class MonthlySummary with _$MonthlySummary {
  const factory MonthlySummary({
    required String month,
    required double income,
    required double expenses,
    required double savings,
    required double net,
  }) = _MonthlySummary;

  factory MonthlySummary.fromJson(Map<String, dynamic> json) => _$MonthlySummaryFromJson(json);
}

@freezed
class CategoryBreakdown with _$CategoryBreakdown {
  const factory CategoryBreakdown({
    String? categoryId,
    required String categoryName,
    required double total,
    required double percentage,
    required String color,
  }) = _CategoryBreakdown;

  factory CategoryBreakdown.fromJson(Map<String, dynamic> json) =>
      _$CategoryBreakdownFromJson(json);
}

@freezed
class MonthlyTrend with _$MonthlyTrend {
  const factory MonthlyTrend({
    required String month,
    required double income,
    required double expenses,
  }) = _MonthlyTrend;

  factory MonthlyTrend.fromJson(Map<String, dynamic> json) => _$MonthlyTrendFromJson(json);
}

@freezed
class ForecastData with _$ForecastData {
  const factory ForecastData({
    required String month,
    required Map<String, dynamic> predictions,
    double? totalPredicted,
    String? confidence,
    String? generatedAt,
  }) = _ForecastData;

  factory ForecastData.fromJson(Map<String, dynamic> json) => _$ForecastDataFromJson(json);
}
